signcrypt_seal.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package saltpack
+
+import (
+	"bytes"
+	"crypto/sha512"
+	"io"
+
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+type signcryptSealStream struct {
+	output        io.Writer
+	encoder       encoder
+	header        *SigncryptionHeader
+	encryptionKey SymmetricKey
+	signingKey    SigningSecretKey
+	keyring       Keyring
+	buffer        bytes.Buffer
+	inblock       []byte
+	headerHash    []byte
+
+	numBlocks encryptionBlockNumber // the lower 64 bits of the nonce
+
+	didHeader bool
+	eof       bool
+	err       error
+}
+
+func (sss *signcryptSealStream) Write(plaintext []byte) (int, error) {
+
+	if sss.err != nil {
+		return 0, sss.err
+	}
+
+	var ret int
+	if ret, sss.err = sss.buffer.Write(plaintext); sss.err != nil {
+		return 0, sss.err
+	}
+	for sss.buffer.Len() >= encryptionBlockSize {
+		sss.err = sss.signcryptBlock()
+		if sss.err != nil {
+			return 0, sss.err
+		}
+	}
+	return ret, nil
+}
+
+func (sss *signcryptSealStream) signcryptBlock() error {
+	var n int
+	var err error
+	n, err = sss.buffer.Read(sss.inblock[:])
+	if err != nil {
+		return err
+	}
+	return sss.signcryptBytes(sss.inblock[0:n])
+}
+
+func (sss *signcryptSealStream) signcryptBytes(b []byte) error {
+
+	if err := sss.numBlocks.check(); err != nil {
+		return err
+	}
+
+	nonce := nonceForChunkSigncryption(sss.numBlocks)
+
+	plaintextHash := sha512.Sum512(b)
+
+	signatureInput := []byte(signatureEncryptedString)
+	signatureInput = append(signatureInput, sss.headerHash...)
+	signatureInput = append(signatureInput, nonce[:]...)
+	signatureInput = append(signatureInput, plaintextHash[:]...)
+
+	detachedSig, err := sss.signingKey.Sign(signatureInput)
+	if err != nil {
+		return err
+	}
+
+	attachedSig := append(detachedSig, b...)
+
+	ciphertext := secretbox.Seal([]byte{}, attachedSig, (*[24]byte)(nonce), (*[32]byte)(&sss.encryptionKey))
+
+	block := []interface{}{ciphertext}
+
+	if err := sss.encoder.Encode(block); err != nil {
+		return err
+	}
+
+	sss.numBlocks++
+	return nil
+}
+
+func receiverEntryForBoxKey(receiverBoxKey BoxPublicKey, ephemeralPriv BoxSecretKey, payloadKey SymmetricKey, index uint64) receiverKeys {
+	// Derive a shared secret by encrypting zeroes, as in the encryption
+	// format. Multiple recipients could potentially claim the same public key
+	// and share this secret.
+	sharedSecretBox := ephemeralPriv.Box(receiverBoxKey, nonceForDerivedSharedKey(), make([]byte, 32))
+	derivedKey, err := rawBoxKeyFromSlice(sharedSecretBox[len(sharedSecretBox)-32 : len(sharedSecretBox)])
```

Comment:
Yes it does. It isn't strictly necessary, because the last step of Poly1305 is to XOR in some stream cipher output. But it's what we did in V1. Will add comments.